### PR TITLE
Snapshot improvements

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
When a snapshot continues in the next epoch, be sure to save the trie nodes in the proper db.
Stop 2 snapshots from starting in parallel for the same rootHash.
Do not close a db until there is another db that has the full state.

### CODE REVIEW

File: trie/trieStorageManager.go [hunk 1]
```diff
@@ -480,18 +522,24 @@ func (tsm *trieStorageManager) takeSnapshot(snapshotEntry *snapshotsQueueEntry,
...
+		log.Error("takeSnapshot: trie storage manager: newSnapshotTrie",
+			"rootHash", snapshotEntry.rootHash,
+			"main trie rootHash", snapshotEntry.mainTrieRootHash,
+			"error", err.Error())
 		return
 	}
 
-	stsm, err := newSnapshotTrieStorageManager(tsm)
+	stsm, err := newSnapshotTrieStorageManager(tsm, snapshotEntry.epoch)
 	if err != nil {
```
Comment (iulianpascalau) [inline trie/trieStorageManager.go:538]: did you forgot to check the error if is of type closing error?
I would change all these "duplicated" error management blocks. Something like:
```GO
    err := do_something_here
    if err != nil{
        treatError(err, "custom message", <roothash>)
        return
    }

//.....
func treatError(err error, message string, rootHash []byte){
    if isClosingError(err){
        log.Debug("context closing", "message", message)
        return
    }

    log.Error(message, "rootHash", rootHash, "err", err.Error())
}

File: trie/trieStorageManagerWithoutPruning_test.go [hunk 2]
```diff
@@ -30,10 +30,10 @@ func TestTrieStorageManagerWithoutPruning_TakeSnapshotShouldWork(t *testing.T) {
 	t.Parallel()
 
 	ts, _ := NewTrieStorageManagerWithoutPruning(testscommon.NewMemDbMock())
-	ts.TakeSnapshot([]byte{}, nil, &trie.MockStatistics{})
+	ts.TakeSnapshot([]byte{}, []byte{}, nil, &trie.MockStatistics{}, 0)
```
Comment (iulianpascalau) [inline trie/trieStorageManagerWithoutPruning_test.go:33]: `make([]byte, 0)` ?

File: storage/pruning/pruningStorer.go [hunk 3]
```diff
@@ -185,24 +183,27 @@ func initPruningStorer(
 	return pdb, nil
 }
 
-func initPersistersInEpoch(
+func (ps *PruningStorer) initPersistersInEpoch(
```
Comment (iulianpascalau) [inline storage/pruning/pruningStorer.go:186]: :+1: 
in another PR, I think we can safely make `PruningStorer` unexported

File: storage/pruning/pruningStorer.go [hunk 4]
```diff
@@ -242,7 +248,13 @@ func initPersistersInEpoch(
 		}
 	}
 
-	return persisters, persistersMapByEpoch, nil
+	ps.activePersisters = persisters
+	ps.persistersMapByEpoch = persistersMapByEpoch
+	return nil
+}
+
+func (ps *PruningStorer) shouldCloseOldPersisters(_ *persisterData) bool {
```
Comment (iulianpascalau) [inline storage/pruning/pruningStorer.go:256]: why do we need this function for? It does not properly overridden by `triePruningStorer.shouldCloseOldPersisters`

File: storage/pruning/pruningStorer.go [hunk 5]
```diff
@@ -807,6 +766,10 @@ func (ps *PruningStorer) extendSavedEpochsIfNeeded(header data.HeaderHandler) bo
 		return true
 	}
 
+	return ps.persisterNeeded()
+}
+
+func (ps *PruningStorer) persisterNeeded() bool {
```
Comment (iulianpascalau) [inline storage/pruning/pruningStorer.go:772]: also here. This is not properly overloaded.

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: common/interface.go [hunk 6]
```diff
@@ -42,8 +42,10 @@ type Trie interface {
 type StorageManager interface {
 	Get(key []byte) ([]byte, error)
 	Put(key []byte, val []byte) error
-	TakeSnapshot([]byte, chan core.KeyValueHolder, SnapshotStatisticsHandler)
-	SetCheckpoint([]byte, chan core.KeyValueHolder, SnapshotStatisticsHandler)
+	PutInEpoch(key []byte, val []byte, epoch uint32) error
+	TakeSnapshot([]byte, []byte, chan core.KeyValueHolder, SnapshotStatisticsHandler, uint32)
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply common/interface.go:46]: Added.

File: storage/pruning/triePruningStorer.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+	defer ps.lock.RUnlock()
+
+	if len(ps.activePersisters) == 0 {
+		return 0, fmt.Errorf("there are no active persisters")
+	}
+
+	return ps.activePersisters[0].epoch, nil
+}
+
+func (ps *triePruningStorer) persisterNeeded() bool {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply storage/pruning/triePruningStorer.go:101]: Refactored.

File: storage/pruning/triePruningStorer.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+		if bytes.Equal(val, []byte(common.ActiveDBVal)) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (ps *triePruningStorer) shouldCloseOldPersisters(pd *persisterData) bool {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply storage/pruning/triePruningStorer.go:116]: Refactored.

File: trie/patriciaMerkleTrie_test.go [hunk 9]
```diff
@@ -437,7 +437,7 @@ func TestPatriciaMerkleTrie_GetSerializedNodesGetFromCheckpoint(t *testing.T) {
 	storageManager := tr.GetStorageManager()
 	dirtyHashes := trie.GetDirtyHashes(tr)
 	storageManager.AddDirtyCheckpointHashes(rootHash, dirtyHashes)
-	storageManager.SetCheckpoint(rootHash, nil, &trieMock.MockStatistics{})
+	storageManager.SetCheckpoint(rootHash, []byte{}, nil, &trieMock.MockStatistics{})
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply trie/patriciaMerkleTrie_test.go:440]: Done.

File: trie/trieStorageManager.go (hunk 1 above)
Comment (BeniaminDrasovean) [reply trie/trieStorageManager.go:538]: There is no need to check for closingError here, because this func does not access the db. 
Refactored as suggested.

File: storage/pruning/pruningStorer.go (hunk 3 above)
Comment (BeniaminDrasovean) [reply storage/pruning/pruningStorer.go:186]: Added TODO.

File: storage/pruning/pruningStorer.go (hunk 4 above)
Comment (BeniaminDrasovean) [reply storage/pruning/pruningStorer.go:256]: Refactored.

File: storage/pruning/pruningStorer.go (hunk 5 above)
Comment (BeniaminDrasovean) [reply storage/pruning/pruningStorer.go:772]: Refactored.

File: storage/pruning/triePruningStorer.go [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+		}
+
+		return val, nil
+	}
+
+	return nil, fmt.Errorf("key %s not found in %s", hex.EncodeToString(key), ps.identifier)
+}
+
+// GetFromLastEpoch searches only the last epoch storer for the given key
+func (ps *triePruningStorer) GetFromLastEpoch(key []byte) ([]byte, error) {
```
Comment (raduchis) [inline storage/pruning/triePruningStorer.go:169]: Last epoch as in not the current, the one exactly before?

File: storage/pruning/fullHistoryTriePruningStorer.go [hunk 11]
```diff
@@ -0,0 +1,56 @@
+package pruning
+
+import (
+	"github.com/ElrondNetwork/elrond-go/storage"
```
Comment (raduchis) [inline storage/pruning/fullHistoryTriePruningStorer.go:4]: imports are not sorted

File: trie/trieStorageManagerWithoutPruning_test.go [hunk 12]
```diff
@@ -30,10 +30,10 @@ func TestTrieStorageManagerWithoutPruning_TakeSnapshotShouldWork(t *testing.T) {
 	t.Parallel()
 
 	ts, _ := NewTrieStorageManagerWithoutPruning(testscommon.NewMemDbMock())
-	ts.TakeSnapshot([]byte{}, nil, &trie.MockStatistics{})
```
Comment (raduchis) [inline trie/trieStorageManagerWithoutPruning_test.go:33]: imports are not sorted

**Review by raduchis — CHANGES_REQUESTED** [review]

File: state/accountsDB.go [hunk 13]
```diff
@@ -1024,20 +1031,35 @@ func (adb *AccountsDB) SnapshotState(rootHash []byte) {
 	defer adb.mutOp.Unlock()
 
 	trieStorageManager := adb.mainTrie.GetStorageManager()
-	if !trieStorageManager.ShouldTakeSnapshot() {
-		log.Debug("skipping snapshot for rootHash", "hash", rootHash)
+	epoch, err := trieStorageManager.GetLatestStorageEpoch()
+	if err != nil {
+		log.Error("SnapshotState error", "err", err.Error())
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply state/accountsDB.go:1036]: Done.

File: common/interface.go [hunk 14]
```diff
@@ -42,8 +42,10 @@ type Trie interface {
 type StorageManager interface {
 	Get(key []byte) ([]byte, error)
 	Put(key []byte, val []byte) error
-	TakeSnapshot([]byte, chan core.KeyValueHolder, SnapshotStatisticsHandler)
-	SetCheckpoint([]byte, chan core.KeyValueHolder, SnapshotStatisticsHandler)
+	PutInEpoch(key []byte, val []byte, epoch uint32) error
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply common/interface.go:45]: This will be removed in another PR, because `increaseNumCheckpoints` is no longer needed. Added TODO.

File: storage/pruning/fullHistoryTriePruningStorer.go [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+		return nil, err
+	}
+
+	ps, err := initPruningStorer(args.StorerArgs, shardId, activePersisters, persistersMapByEpoch)
+	if err != nil {
+		return nil, err
+	}
+
+	tps := &triePruningStorer{ps}
+	ps.extendPersisterLifeHandler = tps.extendPersisterLife
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply storage/pruning/fullHistoryTriePruningStorer.go:38]: Not needed as discussed.

File: storage/pruning/triePruningStorer.go [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+		return nil, err
+	}
+
+	ps, err := initPruningStorer(args, "", activePersisters, persistersMapByEpoch)
+	if err != nil {
+		return nil, err
+	}
+
+	tps := &triePruningStorer{ps}
+	ps.extendPersisterLifeHandler = tps.extendPersisterLife
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply storage/pruning/triePruningStorer.go:33]: Not needed as discussed.

File: storage/pruning/triePruningStorer.go (hunk 10 above)
Comment (BeniaminDrasovean) [reply storage/pruning/triePruningStorer.go:169]: Yes. This is needed because in the case of a trie snapshot, all the values from the previous db need to be added to the current db.

File: storage/pruning/fullHistoryTriePruningStorer.go (hunk 11 above)
Comment (BeniaminDrasovean) [reply storage/pruning/fullHistoryTriePruningStorer.go:4]: Sorted.

File: trie/trieStorageManagerWithoutPruning_test.go (hunk 12 above)
Comment (BeniaminDrasovean) [reply trie/trieStorageManagerWithoutPruning_test.go:33]: Sorted.

File: common/interface.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply common/interface.go:44]: ok

File: storage/pruning/fullHistoryTriePruningStorer.go (hunk 11 above)
Comment (raduchis) [reply storage/pruning/fullHistoryTriePruningStorer.go:6]: ok

File: state/accountsDB.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply state/accountsDB.go]: ok

File: storage/pruning/fullHistoryTriePruningStorer.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply storage/pruning/fullHistoryTriePruningStorer.go:39]: ok

File: storage/pruning/triePruningStorer.go (hunk 10 above)
Comment (raduchis) [reply storage/pruning/triePruningStorer.go:179]: ok

File: storage/pruning/triePruningStorer.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply storage/pruning/triePruningStorer.go:34]: ok

File: trie/trieStorageManagerWithoutPruning_test.go (hunk 12 above)
Comment (raduchis) [reply trie/trieStorageManagerWithoutPruning_test.go:33]: ok

**Review by raduchis — APPROVED** [review]

File: storage/pruning/fullHistoryPruningStorer.go [hunk 17]
```diff
@@ -55,13 +67,13 @@ func initFullHistoryPruningStorer(args *FullHistoryStorerArgs, shardId string) (
 	return fhps, nil
 }
 
-func (fhps *FullHistoryPruningStorer) onEvicted(key interface{}, value interface{}) {
+func (ps *PruningStorer) onEvicted(key interface{}, value interface{}) {
```
Comment (iulianpascalau) [inline storage/pruning/fullHistoryPruningStorer.go:70]: move this in pruningStorer.go ?

File: storage/pruning/triePruningStorer.go [hunk 18]
```diff
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"encoding/hex"
 	"fmt"
+	"github.com/ElrondNetwork/elrond-go/storage"
```
Comment (iulianpascalau) [inline storage/pruning/triePruningStorer.go:7]: go import

File: trie/trieStorageManager_test.go [hunk 19]
```diff
@@ -1,8 +1,11 @@
 package trie_test
 
 import (
+	"github.com/ElrondNetwork/elrond-go-core/core"
```
Comment (iulianpascalau) [inline trie/trieStorageManager_test.go:4]: go imports

File: trie/trieStorageManagerWithoutCheckpoints_test.go [hunk 20]
```diff
@@ -55,11 +55,11 @@ func TestTrieStorageManagerWithoutCheckpoints_SetCheckpoint(t *testing.T) {
 	args := getNewTrieStorageManagerArgs()
 	ts, _ := trie.NewTrieStorageManagerWithoutCheckpoints(args)
 
-	ts.SetCheckpoint([]byte("rootHash"), nil, &trieMock.MockStatistics{})
+	ts.SetCheckpoint([]byte("rootHash"), []byte{}, nil, &trieMock.MockStatistics{})
```
Comment (iulianpascalau) [inline trie/trieStorageManagerWithoutCheckpoints_test.go:58]: `make([]byte, 0)` here and on L62

File: storage/pruning/fullHistoryPruningStorer.go (hunk 17 above)
Comment (BeniaminDrasovean) [reply storage/pruning/fullHistoryPruningStorer.go:70]: Done.

File: storage/pruning/triePruningStorer.go (hunk 18 above)
Comment (BeniaminDrasovean) [reply storage/pruning/triePruningStorer.go:7]: Done.

File: trie/trieStorageManager_test.go (hunk 19 above)
Comment (BeniaminDrasovean) [reply trie/trieStorageManager_test.go:4]: Done.

File: trie/trieStorageManagerWithoutCheckpoints_test.go (hunk 20 above)
Comment (BeniaminDrasovean) [reply trie/trieStorageManagerWithoutCheckpoints_test.go]: Done.

**Review by iulianpascalau — DISMISSED** [review]

File: storage/factory/pruningStorerFactory.go [hunk 21]
```diff
@@ -682,6 +682,22 @@ func (psf *StorageServiceFactory) createTrieEpochRootHashStorerIfNeeded() (stora
 	return trieEpochRootHashStorageUnit, nil
 }
 
+func (psf *StorageServiceFactory) createTriePruningPersister(arg *pruning.StorerArgs) (storage.Storer, error) {
+	isFullArchive := psf.prefsConfig.FullArchive
+	isDBLookupExtenstion := psf.generalConfig.DbLookupExtensions.Enabled
```
Comment (bogdan-rosianu) [inline storage/factory/pruningStorerFactory.go:687]: `isDBLookupExtension` *

File: storage/pruning/triePruningStorer.go [hunk 22]
```diff
@@ -0,0 +1,205 @@
...
+
+	if ps.bloomFilter != nil && !ps.bloomFilter.MayContain(key) {
+		return nil, fmt.Errorf("key %s not found in %s", hex.EncodeToString(key), ps.identifier)
+	}
+
+	if len(ps.activePersisters) < 2 {
+		return nil, fmt.Errorf("key %s not found in %s", hex.EncodeToString(key), ps.identifier)
+	}
+
+	return ps.activePersisters[1].persister.Get(key)
```
Comment (bogdan-rosianu) [inline storage/pruning/triePruningStorer.go:192]: hardcoded values `len(ps.activePersisters) < 2` and `ps.activePersisters[1]`
shouldn't we use the num active persisters from config?

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: storage/pruning/triePruningStorer.go (hunk 22 above)
Comment (BeniaminDrasovean) [reply storage/pruning/triePruningStorer.go]: Refactored, but there is no need for numActivePersisters.

File: storage/pruning/triePruningStorer.go [hunk 23]
```diff
@@ -189,7 +189,8 @@ func (ps *triePruningStorer) GetFromLastEpoch(key []byte) ([]byte, error) {
 		return nil, fmt.Errorf("key %s not found in %s", hex.EncodeToString(key), ps.identifier)
 	}
 
-	return ps.activePersisters[1].persister.Get(key)
+	lastEpochIndex := 1
```
Comment (iulianpascalau) [inline storage/pruning/triePruningStorer.go:192]: could have been a const

**Review by iulianpascalau — APPROVED** [review]

**Review by raduchis — DISMISSED** [review]

File: storage/pruning/triePruningStorer.go (hunk 23 above)
Comment (BeniaminDrasovean) [reply storage/pruning/triePruningStorer.go]: Done.

File: state/accountsDB.go [hunk 24]
```diff
@@ -131,7 +132,25 @@ func NewAccountsDB(
 		},
 		processingMode: processingMode,
 		lastSnapshot:   &snapshotInfo{},
-	}, nil
+	}
+
+	val, err := trieStorageManager.Get([]byte(common.ActiveDBKey))
```
Comment (iulianpascalau) [inline state/accountsDB.go:137]: can we find the `common.ActiveDBKey` in other, old, storers but not in the current one?

Comment (BeniaminDrasovean) [reply state/accountsDB.go:137]: Yes. If the snapshot for the current epoch was started, but then the node was killed, it means that the current db will not have that key.

**Review by raduchis — DISMISSED** [review]

**Review by bogdan-rosianu — DISMISSED** [review]

File: state/accountsDB.go [hunk 25]
```diff
@@ -134,7 +134,7 @@ func NewAccountsDB(
 		lastSnapshot:   &snapshotInfo{},
 	}
 
-	val, err := trieStorageManager.Get([]byte(common.ActiveDBKey))
+	val, err := trieStorageManager.GetFromCurrentEpoch([]byte(common.ActiveDBKey))
```
Comment (iulianpascalau) [inline state/accountsDB.go:137]: :+1:

File: state/accountsDB_test.go [hunk 26]
```diff
@@ -2372,15 +2373,19 @@ func TestAccountsDB_NewAccountsDbStartsSnapshotAfterRestart(t *testing.T) {
 					return true
 				},
 				TakeSnapshotCalled: func(_ []byte, _ []byte, _ chan core.KeyValueHolder, _ common.SnapshotStatisticsHandler, _ uint32) {
+					mutex.Lock()
 					takeSnapshotCalled = true
```
Comment (iulianpascalau) [inline state/accountsDB_test.go:2377]: could have used an atomic.Flag. Can be left as it is.

File: state/peerAccountsDB.go [hunk 27]
```diff
@@ -54,7 +56,25 @@ func NewPeerAccountsDB(
 			},
 			storagePruningManager: storagePruningManager,
 		},
-	}, nil
+	}
+
+	val, err := trieStorageManager.GetFromCurrentEpoch([]byte(common.ActiveDBKey))
```
Comment (iulianpascalau) [inline state/peerAccountsDB.go:61]: since this is duplicate with the AccountsDB implementation, we could have had a commonDB implementation. 
Can be done in another PR (or another commit, if required)

File: trie/trieStorageManager.go [hunk 28]
```diff
@@ -314,6 +314,25 @@ func (tsm *trieStorageManager) Get(key []byte) ([]byte, error) {
 	return tsm.getFromOtherStorers(key)
 }
 
+//GetFromCurrentEpoch checks only the current storer for the given key, and returns it if it is found
+func (tsm *trieStorageManager) GetFromCurrentEpoch(key []byte) ([]byte, error) {
+	tsm.storageOperationMutex.Lock()
```
Comment (iulianpascalau) [inline trie/trieStorageManager.go:319]: I would keep the mutex only to get the storer pointer. 
Otherwise, all GetFromCurrentEpoch will be serialized.

File: storage/pruning/triePruningStorer.go [hunk 29]
```diff
@@ -194,6 +197,22 @@ func (ps *triePruningStorer) GetFromLastEpoch(key []byte) ([]byte, error) {
 	return ps.activePersisters[lastEpochIndex].persister.Get(key)
 }
 
+// GetFromCurrentEpoch searches only the current epoch storer for the given key
+func (ps *triePruningStorer) GetFromCurrentEpoch(key []byte) ([]byte, error) {
+	ps.lock.RLock()
```
Comment (iulianpascalau) [inline storage/pruning/triePruningStorer.go:202]: I would keep the mutex only to get the storer pointer from the map.
Otherwise, all GetFromCurrentEpoch will be serialized.

File: state/peerAccountsDB.go (hunk 27 above)
Comment (BeniaminDrasovean) [reply state/peerAccountsDB.go:61]: Fixed without the commonDB implementation.

File: state/accountsDB_test.go (hunk 26 above)
Comment (BeniaminDrasovean) [reply state/accountsDB_test.go:2377]: Done.

File: trie/trieStorageManager.go (hunk 28 above)
Comment (BeniaminDrasovean) [reply trie/trieStorageManager.go:319]: Done.

File: storage/pruning/triePruningStorer.go (hunk 29 above)
Comment (BeniaminDrasovean) [reply storage/pruning/triePruningStorer.go:202]: Done.

**Review by iulianpascalau — APPROVED** [review]

File: storage/storageUnit/storageunit.go [hunk 30]
```diff
@@ -146,7 +132,8 @@ func (u *Unit) GetOldestEpoch() (uint32, error) {
 
 // Close will close unit
 func (u *Unit) Close() error {
-	//TODO clear the cache and bloom filter
+	u.cacher.Clear()
```
Comment (bogdan-rosianu) [inline storage/storageUnit/storageunit.go:135]: :+1:

File: storage/pruning/triePruningStorer.go [hunk 31]
```diff
@@ -149,7 +149,7 @@ func (ps *triePruningStorer) PutInEpochWithoutCache(key []byte, data []byte, epo
 // GetFromOldEpochsWithoutAddingToCache searches the old epochs for the given key without adding to the cache
 func (ps *triePruningStorer) GetFromOldEpochsWithoutAddingToCache(key []byte) ([]byte, error) {
 	v, ok := ps.cacher.Get(key)
-	if ok {
+	if ok && !bytes.Equal([]byte(common.ActiveDBKey), key) {
```
Comment (raduchis) [inline storage/pruning/triePruningStorer.go:148]: why was this change necessary? It will search for the common.ActiveDBKey in the activePersisters starting with the "oldest" one and returns if it finds one?

**Review by raduchis — APPROVED** [review]
just 2 questions

**Review by LucianMincu — APPROVED** [review]
Initial system tests passed.

More testing will follow in branches.
